### ISSUE TITLE
Cross domain error falsely thrown

### ISSUE THREAD

**Issue by rickharrison** [visionmedia/superagent#484]
I believe that I have been getting some falsely thrown CORs errors.

https://github.com/visionmedia/superagent/blob/master/lib/client.js#L877

I believe that when the page is refreshed or navigated away from while an XHR request is in flight, the browser will set the status to 0. Is there a better way to detect if it is specifically a cross domain error?

**Comment by palamccc** [comment]
+1, One of my xhr request takes 5 seconds to complete, and if page is refreshed before completion, i am getting cross domain violation error. All of my requests are sent to same origin domain only. From stack trace, it looks like the error is generated at https://github.com/visionmedia/superagent/blob/0.21.0/superagent.js#L1280 (Win7, Chrome 40).

It looks like this exception is generated, when page load is aborted by browser. You can see this exception in console only if you enable preserve log in developer console.

**Comment by defunctzombie** [comment]
I am open to suggestions on how to detect this better.

**Comment by defunctzombie** [comment]
Related reading:
http://stackoverflow.com/questions/19325314/how-to-detect-cross-origin-cors-error-vs-other-types-of-errors-for-xmlhttpreq

**Comment by palamccc** [comment]
From what i understand,
When a request is aborted, it may be user interruption (user leaving the page) or due to cross domain violation error. since there is no foolproof way to detect the reason, I think its better to throw generic "Page Aborted" exception. (Developer can anyway find out about cross domain violation from web console.)

As of now code throws self.crossDomainError() for all abort cases.  https://github.com/visionmedia/superagent/blob/0.21.0/superagent.js#L1280

**Comment by zackify** [comment]
this is still happeneing to me. The request goes through, then this error gets thrown. Really weird.

**Comment by grrowl** [comment]
+1 if this is the case it really should be documented. We're capturing many `error.crossDomain = true` errors in Sentry but we've been diagnosing them as if we have a server issue; in fact, it might be regular run-of-the-mill connection aborted during navigation.

As it stands, I don't think we're able to tell at all (and neither is superagent?)

**Comment by ksesong** [comment]
I have the same problem, and cannot find a way around it.
Has anyone found out a way to avoid interrupting the request?

**Comment by kahnjw** [comment]
I am running into this issue as well.

>I am open to suggestions on how to detect this better.

@defunctzombie I think detection is going to be difficult. Right now superagent is gobbling the `xhr` and naively assuming CORs error. That seems wrong. The responsibility of knowing what happened can at least be offloaded to the user, if we give them the `xhr` with `status = 0`. Right now the user has to catch the error and string match on it to do anything about it.

**Comment by kahnjw** [comment]
It should also be trivial to attach some information to the error informing users that it _could_ be a CORs error.

**Comment by AhoyM** [comment]
+1, got this false CORS error when redirecting to another page while requests were in flight.

According to the [wc3 spec][1], The causes of an `xhr.status` of 0 can also be "some type of network error or fetch termination" -  the latter of which applies to my case. As @kahnjw pointed out, we can't assume that a `xhr.status === 0` will be a CORS.

I think a general blanket that covers all these cases is that something went wrong with the outbound request. I'll make a PR that implements that, as well as attaching information on the possible causes. Discussion and suggestions are more than welcome.

[1]: http://www.w3.org/TR/XMLHttpRequest/#the-status-attribute

**Comment by grrowl** [comment]
I think `"Transport Error"` might be more obvious and descriptive than `"Outbound Error"` (I read it as "out of bounds error" for a while)

**Comment by lazreg87** [comment]
+1 Where are we on this?

**Comment by deregtd** [comment]
Latest superagent still throws the same error. :(  I vote for "unknown transport error" or something like that.  Maybe try to be really fancy and detect if the request was a CORS-based request or not (check the url of the source and the target for matching host), and set the crossDomain flag then, but making that the message seems bad.

**Comment by oluckyman** [comment]
As said in #679 that error occurs also if there is no internet connection.

**Comment by rohitlodha** [comment]
+1 @grrowl for transport error. superagent should just pass the error to user.

**Comment by magicdawn** [comment]
Since we can't tell what `status=0` means, we should not swallow the error.
+1 on `Transport Error` & `err.status = 0`

**Comment by jsheetzati** [comment]
+1

**Comment by pornel** [comment]
Agreed that this is misleading and unnecessarily jumping to conclusion.

As mentioned in https://github.com/visionmedia/superagent/pull/653#issuecomment-126088770 fixing this could be a breaking change. As bad as it is, there are applications that check for this particular error.

I suggest keeping API backwards compatible, but extending the message to add something like "or it could be aborted connection, user-agent offline, etc."

**Comment by zackify** [comment]
This has been a problem for so long. I've moved on to using fetch since
it's going to be the new standard for doing Ajax requests.

**Comment by grrowl** [comment]
> I suggest keeping API backwards compatible, but extending the message to add something like "or it could be aborted connection, user-agent offline, etc."

Keep the API backwards compatible, but deploy the changes as a new major version. This is too misleading to keep around in the name of not breaking older apps, and semver exists for a reason.

**Comment by scsper** [comment]
I agree with @grrowl.

**Comment by niftylettuce** [comment]
@zackify @grrowl @scsper @lazreg87 @deregtd @ahoym @kahnjw  @defunctzombie @palamccc @Fishrock123 @rickharrison 

Try out [Frisbee](https://github.com/niftylettuce/frisbee)... it's a replacement I made for this issue we're all having with superagent.

I've documented how to use it specifically with React Native.  See [this React Native pull request](https://github.com/facebook/react-native/pull/4522) I've filed to change the default recommended package to use `frisbee` instead of `superagent` (mainly since `frisbee` uses `fetch` and `superagent` results in the `Origin is not allowed by Access-Control-Allow-Origin` error).
